native/dlahr2.go

```diff
@@ -0,0 +1,143 @@
+// Copyright ©2016 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package native
+
+import (
+	"github.com/gonum/blas"
+	"github.com/gonum/blas/blas64"
+)
+
+// Dlahr2 reduces the first nb columns of a real general n×(n-k+1) matrix A so
+// that elements below the k-th subdiagonal are zero. The reduction is performed
+// by an orthogonal similarity transformation Q^T * A * Q. Dlahr2 returns the
+// matrices V and T which determine Q as a block reflector I - V*T*V^T, and
+// also the matrix Y = A * V * T.
+//
+// The matrix Q is represented as a product of nb elementary reflectors
+//  Q = H_0 * H_1 * ... * H_{nb-1}.
+// Each H_i has the form
+//  H_i = I - tau[i] * v * v^T,
+// where v is a real vector with v[0:i+k-1] = 0 and v[i+k-1] = 1. v[i+k:n] is
+// stored on exit in A[i+k+1:n,i].
+//
+// The elements of the vectors v together form the (n-k+1)×nb matrix
+// V which is needed, with T and Y, to apply the transformation to the
+// unreduced part of the matrix, using an update of the form:
+// A := (I - V*T*V^T) * (A - Y*V^T).
+//
+// The contents of A on exit are illustrated by the following example
+// with n = 7, k = 3 and nb = 2:
+//  [ a   a   a   a   a ]
+//  [ a   a   a   a   a ]
+//  [ a   a   a   a   a ]
+//  [ h   h   a   a   a ]
+//  [ v0  h   a   a   a ]
+//  [ v0  v1  a   a   a ]
+//  [ v0  v1  a   a   a ]
+// where a denotes an element of the original matrix A, h denotes a
+// modified element of the upper Hessenberg matrix H, and vi denotes an
+// element of the vector defining H_i.
+//
+// k is the offset for the reduction. Elements below the k-th subdiagonal in the
+// first nb columns are reduced to zero.
+//
+// nb is the number of columns to be reduced.
+//
+// On entry, a represents the n×(n-k+1) matrix A. On return, the elements on and
+// above the k-th subdiagonal in the first nb columns are overwritten with the
+// corresponding elements of the reduced matrix. The elements below the k-th
+// subdiagonal, with the slice tau, represent the matrix Q as a product of
+// elementary reflectors. The other columns of A are unchanged.
+//
+// tau will contain the scalar factors of the elementary reflectors. It must
+// have length at least nb.
+//
+// t represents the nb×nb upper triangular matrix T and y represents the n×nb
+// matrix Y.
+//
+// Dlahr2 is an internal routine. It is exported for testing purposes.
+func (impl Implementation) Dlahr2(n, k, nb int, a []float64, lda int, tau, t []float64, ldt int, y []float64, ldy int) {
+	checkMatrix(n, n-k+1, a, lda)
```

Comment:
There are no safety checks in the netlib implementation, so I at least added these obvious four. However, not every combination of `n`, `k`, and `nb` is ok, but I'm not sure how strict we are with these very internal routines (it's called only by DGEHRD).